relayer/pkg/chainlink/txm/txm.go

```diff
@@ -0,0 +1,196 @@
+package txm
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/dontpanicdao/caigo"
+	"github.com/dontpanicdao/caigo/types"
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/logger"
+	relaytypes "github.com/smartcontractkit/chainlink-relay/pkg/types"
+	"github.com/smartcontractkit/chainlink-relay/pkg/utils"
+	"github.com/smartcontractkit/chainlink-starknet/relayer/pkg/chainlink/keys"
+	"github.com/smartcontractkit/chainlink-starknet/relayer/pkg/starknet"
+)
+
+const (
+	MaxQueueLen = 1000
+)
+
+type TxManager interface {
+	Enqueue(types.Transaction) error
+}
+
+type StarkTXM interface {
+	relaytypes.Service
+	TxManager
+}
+
+type starktxm struct {
+	starter utils.StartStopOnce
+	lggr    logger.Logger
+	done    sync.WaitGroup
+	stop    chan struct{}
+	queue   chan types.Transaction
+	curve   *caigo.StarkCurve
+	ks      keys.Keystore
+	cfg     starknet.Config
+
+	// TODO: use lazy loaded client
+	client    types.Provider
+	getClient func() types.Provider
+}
+
+func New(lggr logger.Logger, keystore keys.Keystore, cfg starknet.Config, getClient func() types.Provider) (StarkTXM, error) {
+	curve, err := caigo.SC(caigo.WithConstants())
+	if err != nil {
+		return nil, errors.Errorf("failed to build curve: %s", err)
+	}
+	return &starktxm{
+		lggr:      lggr,
+		queue:     make(chan types.Transaction, MaxQueueLen),
+		stop:      make(chan struct{}),
+		getClient: getClient,
+		curve:     &curve,
+		ks:        keystore,
+		cfg:       cfg,
+	}, nil
+}
+
+func (txm *starktxm) Start(ctx context.Context) error {
+	return txm.starter.StartOnce("starktxm", func() error {
+		txm.done.Add(1) // waitgroup: tx sender
+		go txm.run()
+		return nil
+	})
+}
+
+func (txm *starktxm) run() {
+	defer txm.done.Done()
+
+	// TODO: func not available without importing core
+	// ctx, cancel := utils.ContextFromChan(txm.stop)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tick := time.After(0)
+
+	for {
+		select {
+		case <-tick:
+			start := time.Now()
+
+			// calculate total txs to process
+			txLen := len(txm.queue)
+			if txLen > txm.cfg.TxMaxBatchSize() {
+				txLen = txm.cfg.TxMaxBatchSize()
+			}
+
+			// fetch batch and split by sender
+			txsBySender := map[string][]types.Transaction{}
```

Comment:
it's grouped by sender so each sender can have it's own privatekey/signer (CL nodes can have multiple private keys for a key type)

Reply:
I didn't get it, sorry. What's the purpose of that? Why not, let's say, pack the batch up to it's limit every time?

Reply:
hmmm let me try explaining it with more detail :)

the transaction manager is built in a way that it can sign any transaction assuming it has a key for it. each transaction that is placed in the queue has a `senderAddress` parameter. so if i mixed many txs with different senderAddresses into the queue the txm manager should still be able to send all of them - once the txm pulls a batch from the queue it has to split by address so it can use the correct private key to sign those txs.

if it were to pack the batch up to the limit each time there's two problems:
* we want to get txs on chain in a timely manner, so it's unlikely that a batch would become filled
* if we were to put all txs into a single batch (assuming they have mixed senderAddresses), then it would be only using 1 private key to sign the whole batch. and that private key may not be valid for the other senderAddresses

Reply:
Thanks for explanation :)
The batch not being able to contain separately signed txs from different senders - is it our limitation of the tx manager or the network one?

Reply:
I believe it's a network limitation where each "payload" sent to the endpoint has one signature set (r + s components) associated with it, but each "payload" can have multiple txs
- caigo implementation: https://github.com/dontpanicdao/caigo/blob/0a04d64ffe5e6792cef20a53ccc545505cd1ee4b/account.go#L127